src/mono/browser/runtime/diagnostics/client-commands.ts

```diff
@@ -0,0 +1,442 @@
+    Monitoring = 0x200000000,
+    //
+    // Summary:
+    //     Events that will dump PDBs of dynamically generated assemblies to the ETW stream.
+    Codesymbols = 0x400000000,
+    //
+    // Summary:
+    //     Events that provide information about compilation.
+    Compilation = 0x1000000000,
+    //
+    // Summary:
+    //     Diagnostic events for diagnosing compilation and pre-compilation features.
+    CompilationDiagnostic = 0x2000000000,
+    //
+    // Summary:
+    //     Diagnostic events for capturing token information for events that express MethodID
+    MethodDiagnostic = 0x4000000000,
+    //
+    // Summary:
+    //     Diagnostic events for diagnosing issues involving the type loader.
+    TypeDiagnostic = 0x8000000000,
+    //
+    // Summary:
+    //     Events for wait handle waits.
+    WaitHandle = 0x40000000000,
+    //
+    // Summary:
+    //     Recommend default flags (good compromise on verbosity).
+    Default = 0x14C14FCCBD,
+    //
+    // Summary:
+    //     What is needed to get symbols for JIT compiled code.
+    JITSymbols = 0x60098,
+    //
+    // Summary:
+    //     This provides the flags commonly needed to take a heap .NET Heap snapshot with
+    //     ETW.
+    GCHeapSnapshot = 0x1980001
+}
+
+export const enum CommandSetId {
+    Reserved = 0,
+    Dump = 1,
+    EventPipe = 2,
+    Profiler = 3,
+    Process = 4,
+
+    // replies
+    Server = 0xFF,
+}
+
+const enum EventPipeCommandId {
+    StopTracing = 1,
+    CollectTracing = 2,
+    CollectTracing2 = 3,
+    CollectTracing3 = 4,
+    CollectTracing4 = 5,
+}
+
+const enum ProcessCommandId {
+    ProcessInfo = 0,
+    ResumeRuntime = 1,
+    ProcessEnvironment = 2,
+    SetEnvVar = 3,
+    ProcessInfo2 = 4,
+    EnablePerfmap = 5,
+    DisablePerfmap = 6,
+    ApplyStartupHook = 7,
+    ProcessInfo3 = 8,
+}
+
+export const enum ServerCommandId {
+    OK = 0,
+    Error = 0xFF,
+}
+
+function serializeMagic () {
+    return Uint8Array.from(dotnet_IPC_V1);
+}
+
+function serializeUint8 (value:number) {
+    return Uint8Array.from([value]);
+}
+
+function serializeUint16 (value:number) {
+    return new Uint8Array(Uint16Array.from([value]).buffer);
+}
+
+function serializeUint32 (value:number) {
+    return new Uint8Array(Uint32Array.from([value]).buffer);
+}
+
+function serializeUint64 (value:[number, number]) {
+    // value == [hi, lo]
+    return new Uint8Array(Uint32Array.from([value[1], value[0]]).buffer);
+}
+
+function serializeString (value:string|null) {
+    const message = [];
+    if (value === null || value === undefined || value === "") {
+        message.push(...serializeUint32(1));
+        message.push(...serializeUint16(0));
+    } else {
+        const len = value.length;
+        const hasNul = value[len - 1] === "\0";
+        message.push(...serializeUint32(len + (hasNul ? 0 : 1)));
+        for (let i = 0; i < len; i++) {
+            message.push(...serializeUint16(value.charCodeAt(i)));
+        }
```

Comment:
Is there a specific need for UTF-16 here? If you were to use UTF8 you could leverage the TextEncoder APIs directly.

Reply:
Feel free to ignore this comment, I'm only making it on the assumption that this API might be called frequently, if that's not the case (if it's like once per command or something like that, who cares)

Reply:
All those `...array` are very inefficient.

the whole file is optimized for simple code, not for speed. This is manually invoked by hands.